### PR TITLE
Use constructor functions for optional init routines.

### PR DESCRIPTION
**Pull request by sunfishcode**
Instead of using weak symbols, use constructor function attributes for the
environment and preopen initialization routines. This is simpler, uses
less code, and is more LTO-friendly.

### CODE REVIEW

**Comment by kripken** [comment]
Why is this more LTO friendly?

**Comment by sunfishcode** [comment]
It means we don't have to worry about presenting the objects to the linker in a specific order in order to get it to make the correct choice of which definition to use. With constructor functions, there's no choice to make.

File: libc-bottom-half/sources/__environ.c [hunk 1]
```diff
@@ -1,59 +1,63 @@
...
 extern __typeof(__environ) _environ __attribute__((weak, alias("__environ")));
 extern __typeof(__environ) environ __attribute__((weak, alias("__environ")));
 
-// This function is referenced by a weak symbol in crt1.c, and we define
-// it here in the same source file as __environ, so that this function is
-// linked in iff environment variable support is used.
-__wasi_errno_t __wasilibc_populate_environ(void) {
+// We define this function here in the same source file as __environ, so that
+// this function is called in iff environment variable support is used.
+__attribute__((constructor(0)))
```
Comment (sbc100) [inline libc-bottom-half/sources/__environ.c]: I'm not sure about the 0 here.  Presumably we want to reserve some number range for even lower level stuff?  (such as stuff generated by the compiler.. although I guess the linker can put any self-generated stuff before 0 if it chooses).

**Comment by sunfishcode** [comment]
We can discuss whether crt1.o should be a bitcode file when I submit the PR for that ;-).

File: libc-bottom-half/sources/__environ.c (hunk 1 above)
Comment (sunfishcode) [reply libc-bottom-half/sources/__environ.c]: Makes sense. Offhand it's hard to predict what we'd want. The reserved range is 0-100, so I'll set it to 50 to leave room on both sides.

**Pull request merged**